### PR TITLE
refactor: Output ESM files w/ `.mjs` when appropriate

### PR DESCRIPTION
**Pull request by rschristian**
Per the conversation on Slack, both modern and esm output formats may now create `.mjs` files when appropriate.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -121,7 +121,7 @@ export default function (e, t, r) {
 {
 	"main": "./dist/foo.umd.js", // legacy UMD output (for Node & CDN use)
 	"module": "./dist/foo.module.js", // legacy ES Modules output (for bundlers)
```
Comment (rschristian) [inline README.md:123]:
```suggestion
	"module": "./dist/foo.module.mjs", // legacy ES Modules output (for bundlers)
```

Comment (developit) [reply README.md]: whoops, good catch.

File: src/index.js [hunk 2]
```diff
@@ -300,18 +301,22 @@ function getMain({ options, entry, format }) {
 	mainsByFormat.es = replaceName(
 		pkg.module && !pkg.module.match(/src\//)
 			? pkg.module
-			: pkg['jsnext:main'] || 'x.esm.js',
+			: pkg['jsnext:main'] || pkgTypeModule
+			? 'x.esm.js'
+			: 'x.esm.mjs',
```
Comment (developit) [inline src/index.js:306]: I think `.m.js` and `.esm.js` were both poor approximations of `.mjs`, and we should be able to drop the infix.

```suggestion
			: 'x.mjs',
```

File: README.md (hunk 1 above)
Comment (rschristian) [reply README.md]: Nothing like an approval to make mistakes stand out 😅

File: src/index.js (hunk 2 above)
Comment (rschristian) [reply src/index.js:306]: Makes sense, though I wonder if that will confuse some users; for users who might not be super familiar with module formats, Microbundle is quite easy to understand given the current names. `esm` -> `x.esm.js`, `umd` -> `x.umd.js`, etc.

File: src/index.js [hunk 3]
```diff
@@ -439,6 +445,15 @@ function createConfig(options, entry, format, writeMeta) {
...
 
+	// Warn about the (somewhat) breaking change in #950
+	if (format === 'es' && !pkg.module && outputEntryFileName.endsWith('.mjs')) {
+		stdout(
+			yellow(
+				'Warning: If "module" is not specified in your package.json, Microbundle will now output the ESM format with ".mjs" as the file extension in CJS packages. If this is not the desired behavior, please specify "module".',
+			),
+		);
+	}
+
```
Comment (rschristian) [inline src/index.js:456]: New warning message as discussed

File: src/index.js (hunk 2 above)
Comment (developit) [reply src/index.js:306]: It's a reasonable thought, yeah. I could go either way on this.

File: src/index.js [hunk 4]
```diff
@@ -439,6 +445,15 @@ function createConfig(options, entry, format, writeMeta) {
 	const outputDir = dirname(absMain);
 	const outputEntryFileName = basename(absMain);
 
+	// Warn about the (somewhat) breaking change in #950
+	if (format === 'es' && !pkg.module && outputEntryFileName.endsWith('.mjs')) {
+		stdout(
+			yellow(
+				'Warning: If "module" is not specified in your package.json, Microbundle will now output the ESM format with ".mjs" as the file extension in CJS packages. If this is not the desired behavior, please specify "module".',
```
Comment (developit) [inline src/index.js:452]:
```suggestion
				'Warning: your package.json does not specify {"type":"module"}. Microbundle assumes this is a CommonJS package and is generating ES Modules with the ".mjs" file extension.',
```

**Pull request merged**